### ISSUE TITLE
Dart LSP slow startup/first response on large codebase

### ISSUE THREAD

**Issue by RobertBrunhage** [dart-lang/sdk#52947]
`Dart SDK version: 3.0.6 (stable) (Tue Jul 11 18:49:07 2023 +0000) on "macos_arm64"`

# LSP takes a long time before LSP actions work

Not sure how to debug this so if you have tips/approaches, please let me know!

In short, when I open a project with Neovim the dartls LSP starts and in a large project takes towards half a minute for anything related to the LSP to work (hover docs, code completion, code actions etc).

Below I added duration for `dart analyze` with `time dart analyze` with a small project and large project to see the difference. Not sure if this command is even helpful in debugging but at least it shows the difference I experience.

## Small project
```
real    0.65s
user    0.48s
sys     0.35s
cpu     127%
```

## Large project
```
real    31.19s
user    33.99s
sys     8.12s
cpu     135%
```

---

The large project contains, internal packages and is in general very large (>690k LOC counted with `find . -name '*.dart' | xargs wc -l`), but I would still expect it to be faster than 30 seconds for anything to work, and the subsequent action takes ~half the time of that (18 seconds). After those two (first 30 sec wait for a LSP action, and then 18 sec, it becomes almost instant).

Again, if there are better ways to debug this, please let me know.

I will add this here as well as that is where I got the config from (mine is a bit leaner as I don't use `outline` etc.
https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/dartls.lua 

---

Tweet referencing this issue -> https://twitter.com/RobertBrunhage/status/1680840204309913600

**Comment by bwilkerson** [comment]
@DanTup

**Comment by DanTup** [comment]
> Below I added duration for `dart analyze`

Do you mean it takes the same time (~30s) for both `dart analyze` _and_ via the LSP server, or that the LSP server takes ~30s but `dart analyze` is faster?

If you start the server with `--instrumentation-log-file=/path/to/log.txt` it'll produce a log with timestamps that might help understand what's happening during that period (note: the log will likely contain parts of your source code but if you can't share that might be able to use [the timestamps](https://www.epochconverter.com/) at the start of the lines to see where the time is going).

**Comment by RobertBrunhage** [comment]
> > Below I added duration for `dart analyze`
> 
> Do you mean it takes the same time (~30s) for both `dart analyze` _and_ via the LSP server, or that the LSP server takes ~30s but `dart analyze` is faster?

`dart analyze` was just there to have some kind of benchmark for something, and it did showcase about the difference in the perceived LSP speed.

> If you start the server with --instrumentation-log-file=/path/to/log.txt it'll produce a log with timestamps that might help understand what's happening during that period (note: the log will likely contain parts of your source code but if you can't share that might be able to use [the timestamps](https://www.epochconverter.com/) at the start of the lines to see where the time is going).

I tried to add the command but I can't get a log to write it seems. Here is how I setup the LSP
```lua
lsp_config["dartls"].setup({
	on_attach = on_attach,
	cmd = {
		"dart",
		"language-server",
		"--protocol=lsp",
		"instrumentation-log-file=/Users/robertbrunhage/Desktop/lsp-log.txt",
	},
	filetypes = { "dart" },
	init_options = {
		onlyAnalyzeProjectsWithOpenFiles = true,
		suggestFromUnimportedLibraries = true,
		closingLabels = true,
		outline = false,
		flutterOutline = false,
	},
	settings = {
		dart = {
			analysisExcludedFolders = {
				vim.fn.expand("$HOME/AppData/Local/Pub/Cache"),
				vim.fn.expand("$HOME/.pub-cache"),
				vim.fn.expand("/opt/homebrew/"),
				vim.fn.expand("$HOME/tools/flutter/"),
			},
			updateImportsOnRename = true,
			completeFunctionCalls = true,
			showTodos = true,
		},
	},
})
```

LspInfo from NeoVim, but I get no log file

```
1 client(s) attached to this buffer: 
 
 Client: dartls (id: 2, bufnr: [13])
 	filetypes:       dart
 	autostart:       true
 	root directory:  /Users/robertbrunhage/work/mobile
 	cmd:             /Users/robertbrunhage/tools/flutter/bin/dart language-server --protocol=lsp instrumentation-log-file=/Users/robertbrunhage/Desktop/lsp-log.txt
```

---

Still investigating why the log isn't created

**Comment by DanTup** [comment]
> `dart analyze` was just there to have some kind of benchmark for something, and it did showcase about the difference in the perceived LSP speed

Sorry, it's still not clear to me whether `dart analyze` is also taking around the same amount of time, or whether the issue is specific to going through the LSP server. It's expected that the LSP server will be slower to respond during startup and initial analysis, so if the delay you see roughly matches the time taken to run `dart analyze` in the same folder, that's probably expected.


> Got a log now and the first thing I see is the initial request starts at 1689575376779 and the remaining files starts to analyze at 1689575389867, which is 13 seconds

The time prior to the first `publishDiagnostics` includes discovery of all of the projects (which includes walking down the whole open directory looking for `pubspec.yaml`s). Can you confirm whether the folder that is being set as the workspace root is the folder you expect, or if it's something much larger (such as your home directory)? The directory will be in the line that contains `"method"::"initialize"` and could be either `rootPath`, `rootUri` or `workspaceFolders`.

One thing that would probably speed things up at the expensive of only analyzing a subset of projects is setting [`onlyAnalyzeProjectsWithOpenFiles`](https://github.com/dart-lang/sdk/blob/53ead360bba7fe4d8f13ff640967974eb3fce271/pkg/analysis_server/tool/lsp_spec/README.md#initialization-options). Usually the server will analyze the entire folder that the client gives to it, but with this option it will only look for projects related to open files.

**Comment by RobertBrunhage** [comment]
> Sorry, it's still not clear to me whether dart analyze is also taking around the same amount of time, or whether the issue is specific to going through the LSP server. It's expected that the LSP server will be slower to respond during startup and initial analysis, so if the delay you see roughly matches the time taken to run dart analyze in the same folder, that's probably expected.

I would say it roughly matches the time `dart analyze`. So in a small/new project the LSP and dart analyze is pretty much instant <1 sec. On large projects it's >30s for both LSP commands (the first one) and dart analyze. Does that clarify?

> Can you confirm whether the folder that is being set as the workspace root is the folder you expect

This matches the project's folder path (copy pasted and removing the project name: `"rootPath"::"\/Users\/robertbrunhage\/work\/something`)

> One thing that would probably speed things up at the expensive of only analyzing a subset of projects is setting [onlyAnalyzeProjectsWithOpenFiles](https://github.com/dart-lang/sdk/blob/53ead360bba7fe4d8f13ff640967974eb3fce271/pkg/analysis_server/tool/lsp_spec/README.md#initialization-options). Usually the server will analyze the entire folder that the client gives to it, but with this option it will only look for projects related to open files.

I can already see (and did this to try and speed things up) that I am using `onlyAnalyzeProjectsWithOpenFiles` 

`"onlyAnalyzeProjectsWithOpenFiles"::true,"outline"::false,"flutterOutline"::false`

**Comment by DanTup** [comment]
Thanks for the clarification. It sounds like the issue is just that initial analysis takes longer than you'd expect. Without a concrete project to test with, it's hard to say whether the times you're seeing are normal or not.

I'm surprised that `onlyAnalyzerProjectsWithOpenFiles` didn't make a difference though (unless you had files open that touched every project). If you run the server with `--port=8123` and then open http://localhost:8123/ so you see a different number of contexts created with this setting? You should see 0 contexts with no open files, or 1 for each project you have an open file from.

On the left here I had an open workspace with 6 projects, then when I enabled `onlyAnalyzeProjectsWithOpenFiles` and restarted, only the project for the open file produced a context.

![Screenshot 2023-07-24 104542](https://github.com/dart-lang/sdk/assets/1078012/ffc78699-45b3-42b3-bd6c-740c03272152)

If your workspace has a large number of projects and you're not opening files from many of them, I would expect this to make a noticeable difference.

**Comment by RobertBrunhage** [comment]
@DanTup I did the same and is stating `14` contexts. These are all our local packages so I would assume they all get touched because we have barrel files?

![image](https://github.com/dart-lang/sdk/assets/26344867/69dbc094-5376-4bb7-8f8e-2c95e3645dc8)

**Comment by DanTup** [comment]
>  I did the same and is stating 14 contexts

What files did you have open? You shouldn't see lots of contexts when using `onlyAnalyzeProjectsWithOpenFiles` if you don't have files open from 14 projects.

For example, if I open the `pkg` folder in the SDK with `onlyAnalyzeProjectsWithOpenFiles=false` then I see 5` contexts. If I set `onlyAnalyzeProjectsWithOpenFiles=true` and open the same folder, I see 0 contexts. If I open a file (for example `analysis_server\lib\src\lsp\lsp_analysis_server.dart`) then I see a single context.

Can you confirm whether you're seeing 14 contexts with no/one open file, or if you had many files open? If you had zero/one files open, are you able to verify in the log that `onlyAnalyzeProjectsWithOpenFiles` is still `true`, and how many `textDocument.didOpen` lines are logged (to ensure the open files the server is told about match what you're expecting)?

Thanks!

**Comment by joeldegerman99** [comment]
@RobertBrunhage @DanTup I have the same issue where I work. We have around 2.3k files and the analyze and lsp is super slow. I do not have any statistics but if requested I could get some if you want some more data to look at?

**Comment by DanTup** [comment]
@joeldegerman99 it would be interesting to know:

- how many contexts you have (see above)
- how long is "super slow"
- whether this is mainly after you've pulled significant changes to the project or upgraded SDK, or is it also slow if you just run **Developer: Reload Window** in VS Code (eg. everything is fully cached)
- whether you're using `onlyAnalyzeProjectsWithOpenFiles ` (I don't really recommend using this setting - it trades some performance at startup for performance when opening/closing files, but it would be useful to know whether you enabled it)

**Comment by Reprevise** [comment]
@DanTup I have a pretty large Flutter project. It takes ~30 seconds (as seen in the timings) to actually start writing Dart with an active LSP. I only have 1 context, and I use whatever the default is for `onlyAnalyzeProjectsWithOpenFiles`.

Also is it a known issue that the "Analyzer Additional Args" setting doesn't get applied when the **Reload Window** command is triggered?

![image](https://github.com/dart-lang/sdk/assets/25782944/2948c8be-1b1d-4068-8df6-14f380cd7422)

**Comment by DanTup** [comment]
@Reprevise 30s does sound pretty slow to me - does this reproduce after a restart (eg. with no code changes or SDK change)? For comparison, when I open my analysis server workspace, I see:

- 9 contexts
- 3,000 Dart files
- 2.5s for the initialized notification handling

Can you reproduce this on a public project? An instrumentation log may help, but it can contain parts of your source files so is best collected with a public project.

> Also is it a known issue that the "Analyzer Additional Args" setting doesn't get applied when the Reload Window command is triggered?

Nope, it's definitely not :) That command actually tears down the whole extension host and restarts it, so it's not clear to me how Reload Window would behave differently to something like just restarting VS Code. If you enable the [analysis server log](https://dartcode.org/docs/logging/#analyzer) it should record exactly what arguments it's starting the server with. If you think it's not working right, please file an issue at https://github.com/Dart-Code/Dart-Code/issues/new/choose with details. Thanks!

**Comment by Reprevise** [comment]
> Can you reproduce this on a public project? An instrumentation log may help, but it can contain parts of your source files so is best collected with a public project.

I cannot, I tried it with the flutter project and the LSP was pretty speedy, even though its lib folder takes up 3x more space than ours. Perhaps it has something to do with the number of dependencies we use? We have 76 normal dependencies, and 11 dev dependencies.

**Comment by Reprevise** [comment]
> @Reprevise if you copy your list of dependencies/dev_dependencies into a new project (and add imports for them), does it repro?

Do the imports have to be used?

**Comment by DanTup** [comment]
@joeldegerman99  I wouldn't expect private packages to be any different to public packages. Are they just `path` packages pointing to other folders? You could try commenting them out of your pubspec (and running pub get) to see if whether that changes the timings disproportionately.

@Reprevise I'm not certain, but I don't think so - I think the analyzer just needs to "discover" the files to load them. However if it seems to not work let me know and I'll do some testing. 

Some other numbers that could be useful to know.. if you run the **Dart: Open Analyzer Diagnostics** from VS Code and then click on the "Contexts" page, for each context under "Context Files" there will be a count of "added files" and "implicit files". Those counts may be useful (for each context) to get an idea of how many files there are including dependencies.

![image](https://github.com/dart-lang/sdk/assets/1078012/071c0009-4d65-41dc-88b7-dd1aae86ee6d)

**Comment by DanTup** [comment]
@Reprevise that's odd - are there any differences in things that might not be in the source folder (for example VS Code settings that might be gitignored) that could be different between them?

**Comment by DanTup** [comment]
Hmm, that's interesting to know - it's probably too late now if you've already cleaned it, but it'd be interesting to know why it got so big. @joeldegerman99 it might be interesting for you to try the same - rename the `.dart_tool` folder (so we can poke around in it if it helps), re-run `pub get` for all the projects, then time it again.

**Comment by DanTup** [comment]
@Reprevise out of interest, does your project have any code-gen? If you run your app a few times (maybe making some trivial changes) , do any files start building up in that directory?

**Comment by Reprevise** [comment]
> @Reprevise out of interest, does your project have any code-gen? If you run your app a few times (maybe making some trivial changes) , do any files start building up in that directory?

Yes, our project uses code-gen pretty heavily. On the first run of build_runner, the .dart_tool folder went from 89.1 MB (I did run the app after running `flutter clean`) to 148 MB. After adding and then removing fields of a freezed class, it went to ~151 MB.

**Comment by DanTup** [comment]
@Reprevise thanks, I'll do some testing!

**Comment by DanTup** [comment]
@Reprevise I did some testing with the Freezed example, and although the `.dart_tool` folder grow initially, it didn't continue to grow as I kept making changes (it stuck at 50.6MB on my machine for the example app). I tried different freezed versions and changing SDK constraints, but it didn't seem like anything caused orphaned files or the folder to keep growing.

If it does keep growing for you, please let me know the steps.

**Comment by DanTup** [comment]
@Reprevise @joeldegerman99 also - could you confirm which OSes you're on? We have some thoughts about the time spent setting up watchers, and the watcher implementations are platform-specific. Thanks!

**Comment by Reprevise** [comment]
> @Reprevise @joeldegerman99 also - could you confirm which OSes you're on? We have some thoughts about the time spent setting up watchers, and the watcher implementations are platform-specific. Thanks!

I'm on Windows 11. I'll get back to you on your previous comment tonight.

**Comment by DanTup** [comment]
@Reprevise I also forgot to ask - what's the disk/filesystem these files are on? Is it a local filesystem (not remote), and is it an SSD or mechanical drive?

**Comment by Reprevise** [comment]
> @Reprevise I also forgot to ask - what's the disk/filesystem these files are on? Is it a local filesystem (not remote), and is it an SSD or mechanical drive?

Local filesystem, and I use SSDs. NTFS is the filesystem.

**Comment by joeldegerman99** [comment]
> @Reprevise @joeldegerman99 also - could you confirm which OSes you're on? We have some thoughts about the time spent setting up watchers, and the watcher implementations are platform-specific. Thanks!

Windows 11 😊

**Comment by DanTup** [comment]
@joeldegerman99 

> My .dart_tool is 3.2gb and I have created a copy of it. Do you want me to give it to you?

Did renaming the folder away speed up the analysis? If not, it's perhaps unrelated - although I would be interested in a directory listing to understand what kind of contents are in that folder. You can generate a directory listing by running the following command in a PowerShell terminal from the large copy of the `.dart_tool` folder:

```
Get-ChildItem -Recurse -Force -ErrorAction SilentlyContinue | Select-Object FullName, Length | Out-File "dir-listing.txt"
```

Please check there's nothing sensitive in the file before sharing. In particular I'm interested in knowing whether there are lots of files/folders or just a small number of large ones, and how many .dart files and/or pubspec.yaml files there might be (which might be analyzed).

Thanks!

**Comment by joeldegerman99** [comment]
> @joeldegerman99
> 
> > My .dart_tool is 3.2gb and I have created a copy of it. Do you want me to give it to you?
> 
> Did renaming the folder away speed up the analysis? If not, it's perhaps unrelated - although I would be interested in a directory listing to understand what kind of contents are in that folder. You can generate a directory listing by running the following command in a PowerShell terminal from the large copy of the `.dart_tool` folder:
> 
> ```
> Get-ChildItem -Recurse -Force -ErrorAction SilentlyContinue | Select-Object FullName, Length | Out-File "dir-listing.txt"
> ```
> 
> Please check there's nothing sensitive in the file before sharing. In particular I'm interested in knowing whether there are lots of files/folders or just a small number of large ones, and how many .dart files and/or pubspec.yaml files there might be (which might be analyzed).
> 
> Thanks!

From what I can see in the dir listing is that there is not sensitive data in that file. There is a lot of kernel snapshots and generated files

[dir-listing.txt](https://github.com/dart-lang/sdk/files/13631797/dir-listing.txt)

**Comment by DanTup** [comment]
@joeldegerman99 thanks - can you confirm whether renaming the folder actually made the analysis faster? It doesn't seem like there's a huge number of files here (and as far as I can tell, no Dart files - although some of the lines are truncated).

@christopherfujino is it expected that `.dart_tool\flutter_build` will grow over time unless the user cleans it, or should something be pruning it?

```
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\.filecache                 
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\app.dill                   
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\debug_android_application.…
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\flutter_assets.d           
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\gen_dart_plugin_registrant…
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\gen_localizations.stamp    
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\kernel_snapshot.d          
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\kernel_snapshot.stamp      
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e2ff41db772a6058ec43be2e4dcee408\outputs.json               
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\.filecache                 
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\app.dill                   
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\debug_android_application.…
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\flutter_assets.d           
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\gen_dart_plugin_registrant…
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\gen_localizations.stamp    
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\kernel_snapshot.d          
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\kernel_snapshot.stamp      
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\e3ed0387e6243e10ee1daa57fbd354c5\outputs.json               
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\.filecache                 
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\app.dill                   
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\debug_android_application.…
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\flutter_assets.d           
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\gen_dart_plugin_registrant…
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\gen_localizations.stamp    
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\kernel_snapshot.d          
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\kernel_snapshot.stamp      
C:\Users\JoelDegerman\Desktop\.dart_tool_copy\flutter_build\f0d011fae7cb96c3ea997ca8ffec48af\outputs.json   
```

**Comment by DanTup** [comment]
@joeldegerman99 

It also occurs to me that renaming the folder (but not moving it outside of the project) might not help, because if it's slow to enumerate/analyze, it's still in the workspace. So you'd need to move it away.

**Comment by Reprevise** [comment]
> If it does keep growing for you, please let me know the steps.

Sorry for the late response. I got mixed results when changing the fields of a freezed class. Sometimes it increased a little bit, sometimes it stayed the same. I think it _has_ to be something in the `.dart_tool` folder because once I cleaned the project the analysis was instantly sped up and we know it balloons over time. It just makes sense, but I could also be totally off base.

**Comment by joeldegerman99** [comment]
> > @christopherfujino is it expected that .dart_tool\flutter_build will grow over time unless the user cleans it, or should something be pruning it?
> 
> I don't believe we ever garbage collect this, but also I would be surprised if this was slowing down analysis initialization. If so, I wonder if it would be safe for the analyzer to hardcode `.dart_tool/flutter_build` to be a path to ignore.
> 
> > @joeldegerman99 thanks - can you confirm whether renaming the folder actually made the analysis faster?
> 
> I am also curious about this. In addition, do you know if you have any virus-scan operating on C:\Users\JoelDegerman\Desktop ? If so, then just listing the directory contents could potentially be slow if it triggers a scan first.

No virus-scan operations is run except windows defender.

**Comment by joeldegerman99** [comment]
> @joeldegerman99
> 
> It also occurs to me that renaming the folder (but not moving it outside of the project) might not help, because if it's slow to enumerate/analyze, it's still in the workspace. So you'd need to move it away.

Yeah I moved it out the first thing I did and saved a copy.

**Comment by joeldegerman99** [comment]
My analyze speed has not been improved. My biggest issue is that it is not consistently slow or fast it is a mix of both and annoying. My laptop performance is not the issue since it is on the upper end of better performance.
![image](https://github.com/dart-lang/sdk/assets/79711461/1ca147ea-2793-4219-a6f3-9a9c501b6b4f)

**Comment by christopherfujino** [comment]
> No virus-scan operations is run except windows defender.

Defender would count as a virus scanner.

**Comment by DanTup** [comment]
Thanks for the info - I have a few things to look at that might help here:

1. Duplication of file watchers during initial analysis
  We currently create an additional watcher for each workspace folder during setup. On Windows, the directory watchers involves a [full recursive walk](https://github.com/dart-lang/watcher/blob/dc45f1925f66743ff29b9dcbb4693a7ec25945c5/lib/src/directory_watcher/windows.dart#L411) of the directory, so any time spent here (which scales with the number of files) could be done twice
2. Overlapping of watchers
  If you have nested folders, it's possible that we create watchers that overlap (eg. one for the parent project and one for the child project). This may also cause duplicate walking down the tree (and goes up with the number of nested projects)
3. Improve the timings so you can click on the Initialized request and see a clearer breakdown of what time was spent discovering projects, setting up watchers, etc.)

@joeldegerman99 since clearing `.dart_tool` didn't help, I wonder how big your project folder is generally (the whole thing, not just `.dart_tool`). Do you have any other large folders in there (for example if you had a nodejs project with a huge node_modules folder, we'd probably be walking that too).

And do you know if there's any pattern to it being slow/fast? Is it slow usually after you've changed SDK and/or pulled other code changes from git?

**Comment by longtimedeveloper** [comment]
@DanTup do you have a feature we can enable that will log what LSP is doing so you can get outstanding information to analyze?

**Comment by Reprevise** [comment]
Hi @DanTup!

Just wanted to reply back here and the issue is present again. The trick to speed it up for me is to remove my 2GB `build` folder and (~460MB) `.dart_tool` folder from my project (normally I'd delete them but I've saved them just incase you want to take a peek).

**Comment by longtimedeveloper** [comment]
@DanTup @Reprevise why would the LSP scan the `build` or `.dart_tool` folder?  

I know that I do not know all of your requirements.  Is there a way for developers to blacklist folders so that the tool will not scan them?  Thank you.  Happy New Year.

**Comment by DanTup** [comment]
@longtimedeveloper 

> Dan, just just read this suggestion and instruction from above, --instrumentation-log-file=/path/to/log.txt
> 
> Please give me the full command you need me to run. I have not idea how to start LSP from the command line.

In VS Code you can set the `"dart.analyzerInstrumentationLogFile"` setting, for ex:

```
"dart.analyzerInstrumentationLogFile": "/tmp/instrum.txt",
```

> why would the LSP scan the `build` or `.dart_tool` folder?

There's nothing preventing you having a folder named `build` in your project. While it might be unlikely there's code in there, I don't know if it would be obvious that this folder would be excluded by default. If it turns out to be significantly harming performance, perhaps it should be considered - but it would be better to have some more concrete data on this.

`.dart_tool` can also contain packages like `flutter_gen` (a generated package used for localization) so it needs to be analyzed like any other part of your project.

> Is there a way for developers to blacklist folders so that the tool will not scan them?

In VS Code, you can use the `dart.analysisExcludedFolders` setting to exclude folders from analysis, however I don't think it would prevent `package:watcher` from descending into them (though there's an issue at https://github.com/dart-lang/sdk/issues/54274 to explore moving off package:watcher which might provide more gains than just reusing some of the watchers as noted above).


@Reprevise 

> The trick to speed it up for me is to remove my 2GB build folder and (~460MB) .dart_tool folder from my project

I'd be interested in knowing how many files/folders there are in them (if you're on Windows, there's a command above at https://github.com/dart-lang/sdk/issues/52947#issuecomment-1846930553 that would make a recursive directory listing). A small number of large files shouldn't be too bad, but a large number of small items could make the watcher creation slower.


I had an idea to create a script that creates watchers on these folders and records the times it takes. This could allow us to get more detailed times than the server currently records without having to ship changes in the SDK (SDK round trip time is pretty slow). Using `package:analyzer` we might also be able to run the code that locates the contexts in the project, so we could also time that. This might help verify if the problem is the watchers, and if so what the impact of some of the changes noted above could be.

I'll ping back here if I can get something working that you could try out.

**Comment by Reprevise** [comment]
@DanTup I have sent an email to logs@dartcode.org with the directory listings of my old `.dart_tool` and `build` folders.

**Comment by longtimedeveloper** [comment]
@DanTup this is my daily routine and it works for me.

1. Open my C# Azure project in VSCode.  Wait until LSP failed to start message appears.
2. Close VSCode
3. Reopen my C# Azure project in VSCode and everything works within 30 seconds.

4. Start Azurite

5. Open my Flutter shared app in VSCode.  Wait 5 minutes and then everything will work.

6. Open my Flutter desktop app in VSCode.  Wait 5 minutes and then everything will work.

Rather than get frustrated, much easier to have a coffee and do the above routine.  I am actually OK with this since once I get going, everything works for the next 8-14 hours of work I do each day.

I will send you log files from an empty project soon.  The last few days here in Ukraine and in my city have been horrible.

**Comment by DanTup** [comment]
@longtimedeveloper

> Wait until LSP failed to start message appears.

Can you clarify what this message is? This issue has been about performance but I wasn't aware of any error messages being shown.

**Comment by Salakar** [comment]
👋 hope I'm not polluting this thread with a drive-by comment. For zapp.run (which has the Dart analyzer in LSP mode compiled into a JS web worker) we had this issue (but on any code base) and it'd take up to a minute for initialized to happen. 

I thought it was due to the nature of it being compiled to JS but I investigated to see if there's anything I could tweak for performance and found this: https://github.com/dart-lang/sdk/blob/main/pkg/analyzer/lib/src/dart/analysis/driver.dart#L2078

After patching it to a much lower `4` we had startup happen in roughly less than a third of the original time (and after some other patches which I don't think are relevant here we now have instant analyzer start as you can see on the live site). We've had no other visible side-effects of doing this patch for Zapp and the code comment above even indicates it's an arbitrary number.

**Comment by DanTup** [comment]
@Reprevise thank you for the log! I see that you had around 54,000 files in your `build` folder - a lot of which seemed to come from plugins.

I created a new project with the same plugin list and built the app in debug, profile and release mode. This resulted in 34,000 files in my build folder. Initial startup takes around 12seconds on my machine, which although not as bad as you see (possibly because I'm on fast NVMe drives) should at least give me something reasonable to profile that might be indicative if what you're seeing.

I'll do some testing and post back what I find.

@Salakar that's interesting to know. I'll see what impact changing it has running locally, though I suspect optimal values could vary between environments (for example the IO might be different between disk access and the APIs you're using).

**Comment by Reprevise** [comment]
> possibly because I'm on fast NVMe drives

Hm, I'm also on particularly fast M.2 NVMe drives (Samsung 980 PRO). With a ~1.3GB `build` folder (created two days ago) and a ~140MB `.dart_tool` folder, the LSP has the `initialized` timing at ~9.3s. That number isn't on a cold start of VSCode, just after a Reload Window command. On a true cold start (after a pc reboot), it's about ~30s but that could be many other factors.

**Comment by DanTup** [comment]
Ah, that's interesting to know. Seems like our timings may be similar then (9s vs 12s). I found some things that may be contributing to that and once I'm done investigating that I'll see if I can repro your much slower times after a reboot (but I agree there may be other factors contributing to that which we might not be able to influence, since there's no real difference for us between a reboot and Reload Window - you're getting a fresh analysis server in both cases).

**Comment by longtimedeveloper** [comment]
> @longtimedeveloper
> 
> > Wait until LSP failed to start message appears.
> 
> Can you clarify what this message is? This issue has been about performance but I wasn't aware of any error messages being shown.

@DanTup 

Yes, I can send you a screenshot tomorrow after I start my computer, no problem.

**Comment by DanTup** [comment]
> the LSP has the initialized timing at ~9.3s. That number isn't on a cold start of VSCode, just after a Reload Window command. On a true cold start (after a pc reboot), it's about ~30s but that could be many other factors.

Seems I can reproduce this. After a reboot, I saw 25-30 seconds for the first time I opened VS Code and looked at the timing, and also for the second. On the third, it was 12s.

It's possible there's some caching going on (in Windows/the disk) that's affecting this - although it's also possible that using the time from the Initialized notification is misleading (because it has `await`s and could be delayed by unrelated work that starts in the background depending on timing). I'll do some profiling to see what I can find anyway. Thanks!

**Comment by longtimedeveloper** [comment]
@DanTup 

LSP Start Up Error I get the first time each day that I open a C# project in VSCode

`Using dotnet configured on PATH
Dotnet path: C:\Program Files\dotnet\dotnet.exe
Activating C# + C# Dev Kit + C# IntelliCode...
waiting for named pipe information from server...
[Error - 5:45:33 AM] Microsoft.CodeAnalysis.LanguageServer client: couldn't create connection to server.
Error: Timeout. Named pipe information not received from server.
	at q.<anonymous> (c:\Users\Karl\.vscode\extensions\ms-dotnettools.csharp-2.14.8-win32-x64\dist\extension.js:2:1290527)
	at Generator.next (<anonymous>)
	at s (c:\Users\Karl\.vscode\extensions\ms-dotnettools.csharp-2.14.8-win32-x64\dist\extension.js:2:1280557)
[stdout] {"pipeName":"\\\\.\\pipe\\20aef8e7"}
received named pipe information from server`


![lspstartuperror](https://github.com/dart-lang/sdk/assets/79998776/54dd0c67-97b8-40e6-98d4-1b231a365be0)

**Comment by Reprevise** [comment]
@longtimedeveloper Unless I'm missing something, that appears unrelated to Dart whatsoever.

**Comment by DanTup** [comment]
@longtimedeveloper yeah, that looks like an error from the C# extension and (unless we've done something _really_ bad) I don't think we're involved in that one.

A small update - I've landed some changes (see https://github.com/dart-lang/sdk/issues/54513) that avoid some redundant work at startup. This reduces the time from sending the initialize request to getting the end of first analysis (measured by a script) from ~9s to ~5.3s for my build folder with 33k files (and a similar proportion increase for the cold start).

The longer cold start seems to be that file access is slower the first (few) times which I suspect is because less there's nothing cached by the OS compared to subsequent opens. This might be trickier to improve, but I'll do some more digging.

**Comment by caibou** [comment]
> > I did the same and is stating 14 contexts
> 
> What files did you have open? You shouldn't see lots of contexts when using `onlyAnalyzeProjectsWithOpenFiles` if you don't have files open from 14 projects.
> 
> For example, if I open the `pkg` folder in the SDK with `onlyAnalyzeProjectsWithOpenFiles=false` then I see 5`contexts. If I set`onlyAnalyzeProjectsWithOpenFiles=true`and open the same folder, I see 0 contexts. If I open a file (for example`analysis_server\lib\src\lsp\lsp_analysis_server.dart`) then I see a single context.
> 
> Can you confirm whether you're seeing 14 contexts with no/one open file, or if you had many files open? If you had zero/one files open, are you able to verify in the log that `onlyAnalyzeProjectsWithOpenFiles` is still `true`, and how many `textDocument.didOpen` lines are logged (to ensure the open files the server is told about match what you're expecting)?
> 
> Thanks!

Is this related to the fact that we use multiple modules in the project?
![Screenshot 2024-04-30 at 08 54 22](https://github.com/dart-lang/sdk/assets/19165682/bfed6e95-4c57-4ba3-bc11-d2619b129dfb)

**Comment by DanTup** [comment]
@caibou having multiple projects (I'm assuming that's what you mean by modules?) in the workspace will increase analysis time, but that shouldn't itself be a problem (it's very common to have multiple projects in a workspace).

**Comment by DanTup** [comment]
The fixes I noted further up shipped in the latest versions of Dart/Flutter released last week.

Looking back at the original report here from @RobertBrunhage I see that `dart analyze` took just over 30s. This looks suspicious and I suspect this may have been https://github.com/dart-lang/sdk/issues/55621 - which was also fixed in the latest SDK (via https://github.com/dart-lang/sdk/issues/55402).

I'm interesting in hearing your experiences after upgrading to the latest SDKs (Flutter 3.22 and Dart 3.4).

**Comment by Reprevise** [comment]
I'm going to be testing out Flutter 3.22 today on our large app to be sure it's ready for production use so I'll get an insight into the analyzer performance soon.

**Comment by Reprevise** [comment]
I'm seeing 15s long initialization times w/o any analyzer plugins. I had `custom_lint` as a plugin to analyze riverpod stuff which made initialization take ~45s so I've removed that for now.

Still though, taking out my ~5.8GB build folder and ~500MB .dart_tool folder made it so much faster (~4 seconds instead of ~15).

**Comment by DanTup** [comment]
> I'm seeing 15s long initialization times w/o any analyzer plugins. I had `custom_lint` as a plugin to analyze riverpod stuff which made initialization take ~45s so I've removed that for now.

Are you able to capture an instrumentation log with the plugin enabled? It adding 30s sounds suspicious, but the fix for that should've been included in 3.22.

What's the timing (without plugins) if you take out only the `build` folder and not `.dart_tool` (eg., if `build` was somehow excluded by default for Flutter apps, where would that leave you)? And does adding `"dart.analysisExcludedFolders": ["build"]` to your settings also get you to that same time?

**Comment by Reprevise** [comment]
I'll send the instrumentation log w/ custom_lint to the logs email shortly.

Having the bloated `.dart_tool` folder makes no noticeable impact to the timing.

With regards to deleting vs excluding the build folder, deleting the folder feels slightly faster than excluding it though I have no solid evidence to back that up.

**Comment by DanTup** [comment]
@Reprevise thanks! Looking at the log you sent me, it appears that the server started at 15:49:04.926 (GMT) and the first analysis ended at 15:49:21.133 which looks like around 15 seconds (and the initialization appears to complete very quickly). I can also see the plugin generating diagnostics from 15:49:37.494 to 15:49:48.198 which is approx 11 seconds. What is it you're measuring when you're seeing 45s? (as far as I can tell, you're not hitting the 30s issue I noted should've been fixed in this version).

> With regards to deleting vs excluding the build folder, deleting the folder feels slightly faster than excluding it though I have no solid evidence to back that up.

If you're able to capture a log of each of these (without the plugin) I'd be interested in comparing them. I hoped there would be no difference, but I don't know if excluding prevents us walking the folder entirely.

**Comment by Reprevise** [comment]
@DanTup I'm going to send you another log but it'll be unrelated to your request about the `build` folder. I've captured the long startup time and what I believe to possibly be another issue where after not doing anything in VSCode then coming back, the analysis server re-analyzing the whole project again.

About the long startup time, would it be correct to measure the time between the first entry in the log, and the entry right before the first publishDiagnostics?

**Comment by DanTup** [comment]
> where after not doing anything in VSCode then coming back, the analysis server starts re-analyzing the whole project again.

That does sound odd - it's probably worth filing a specific issue about that with details (and I'll review the log you send). If you can include approx timestamps of when you first opened the project, went away, and came back, that may be useful too.


> About the long startup time, would it be correct to measure the time between the first entry in the log, and the entry right before the first publishDiagnostics?

When trying to understand if something is a startup/initial analysis issue, I'll look for when this appears in the log:

```
Res:{"jsonrpc"::"2.0","method"::"$/progress","params"::{"token"::"ANALYZING","value"::{"kind"::"end"}}}
```

This is when the server believes it has completed analysis. However, the specifics of what is startup/initial analysis is probably not so important here if _something_ appears to be taking much longer when you open a project. If it seems to you that something at startup is taking ~45s then I'm interested in understanding what that might be (but that requires understanding what you're seeing to reach that conclusion - what are you seeing/measuring?).

**Comment by Reprevise** [comment]
@DanTup I've opened a new issue and included the line in the log when I believe I came back.

To summarize (because there are so many comments here):
- Large `build` folder adds a lot of time to the LSP (this is the main issue, which is solved by either periodically deleting the build folder, or excluding it from analysis in VSCode settings)
- Plugins like `custom_lint` add ~30s to LSP "initialization" (but its still usable before this initialization w/ features like autocomplete, etc)

This slow startup time (~15s) is my main concern here, I personally don't really care about the ~30s of `custom_lint` startup because the LSP is still usable.

**Comment by DanTup** [comment]
@Reprevise 

> Plugins like custom_lint add ~30s to LSP "initialization"

Can you confirm what you're measuring for this? Do you mean it takes an extra 30s for custom_lint to report its own errors, or something else? (I know it's not your priority, but if plugins are causing any non-plugin functionality from the server to be delayed by anything like 30s, I would like to understand what's happening there).

Will think about how we might be able to handle `build` without just globally excluding it (there's no reason you couldn't have a `build` folder in a non-Flutter project, or nested inside your project that you might expect to be analyzed).

**Comment by Reprevise** [comment]
@DanTup

> Can you confirm what you're measuring for this? Do you mean it takes an extra 30s for custom_lint to report its own errors, or something else? (I know it's not your priority, but if plugins are causing any non-plugin functionality from the server to be delayed by anything like 30s, I would like to understand what's happening there).

Sure, so what I mean is that the LSP starts up, and reports the 18 diagnostics in my project, and autocomplete and the rest of the LSP features come online. Then, `custom_lint` takes ~30s to report all its diagnostics and that's when I'm guessing the "initialized" event gets sent.

**Comment by DanTup** [comment]
> Then, `custom_lint` takes ~30s to report all its diagnostics and that's when I'm guessing the "initialized" event gets sent.

The `initialized` event is actually sent by the client to the server. The timing shown in the diagnostic page is how long the server took to execute all code in the handler for that event. At a glance, I can't see that it should wait on anything like analysis or plugins - is it possible you can send me a matching log file and a screenshot of  the Timing page? If possible, close all open files and restart VS Code then open the project but don't open any files until > 30s when you think it's finished (this will keep the log clearer of unrelated noise).

Could you also confirm the exact version of `custom_lint`, `analyzer`, `analyzer_plugin` you're using? (from your pubspec.lock or .dart_tool/package_config.json`) for me to see if I can reproduce locally.

Thanks!

**Comment by Reprevise** [comment]
@DanTup

I have emailed you a clean instr log of the analyzer, w/o me opening any files.

The versions are as follows:
- analyzer: 6.4.1
- analyzer_plugin: 0.11.3
- custom_lint: 0.6.4

**Comment by DanTup** [comment]
@Reprevise sorry for the delay - I just looked at this but you didn't include a copy of the timing page showing exactly how long the initialized event was recorded as. Could you capture a log file and include a matching screenshot showing the exact timing recording for the initialized event when the plugin is enabled (but, without the large build folder).

(I could understand the large build folder making `initialized` take longer because the server would walk that folder, but I'm trying to track down why plugins would be doing it - in my testing I'm getting quite fast `initialized` times).

**Comment by Reprevise** [comment]
@DanTup I have emailed you the logs and the screenshot of the timings page. The log is w/o the large build folder but does have `custom_lint` enabled.

Incase you can't see the screenshot in the email, here it is:
![dart lsp timings page](https://github.com/dart-lang/sdk/assets/25782944/c530229f-11dc-4eda-aa72-3d4756cd46e7)

**Comment by PaulGrandperrin** [comment]
Hi,
I just want to share my experience, since it seems that it's a little bit on the extreme side.

I'm using vscode on a small/medium sized flutter app.

The dart language server takes about **half an hour to one hour** before it can offer any completion or type information to the IDE.

While I'm waiting for it to start working, it uses about 65% of 1 CPU core (when I have 4 cores / 8 threads) and basically does almost no IOs.

![Image](https://github.com/user-attachments/assets/71d19ade-1927-4aed-84b6-4d61ec058bf8)

Since an hour is a long time to wait every time I launch vscode, I tried many things to understand what it is doing and why it's not even 100%ing my CPU or my SSD.

`sysdig -c topfiles_times.lua` didn't show anything useful, nor `strace -f -e trace=open,openat -p $DART_PID`.

Finally, I found a lead with `ls -l /proc/$DART_PID/fd/`. Here I can see that many files are open in random places in my `/nix/store/` (I'm using NixOS).

So it seems that it is scanning my whole OS, which is huge, and doing so very slowly to avoid saturating the CPU or the SSD.

Looking more at the opened file descriptors, I see that the `.direnv` folder from my project folder. `direnv` is a small utility to have per-project dev environments. My direnv points to files in my `/nix/store/` where my dev env is stored. I guess that's how it then looses itself scanning my whole `/nix/store/`.

So I'm trying to exclude `.direnv` from the dart analysis, but it seems this parameter is not working, or I am missing something because nothing changes:
```
 "dart.analysisExcludedFolders": [
    ".direnv",
  ],
```

**Comment by DanTup** [comment]
@PaulGrandperrin could you file a specific issue for the problem you're seeing (that is, excluded folders are still enumerated during startup) and CC me?

**Comment by PaulGrandperrin** [comment]
@DanTup Ok I'll do that when I go back to this problem and investigate a little more.

@bwilkerson Yes, the way `direnv` and `nix` work together is by creating a tree of symlinks in the `.direnv` folder that points to packages in `/nix/store` and this nix store is also a huge tree of symlinks!

So my guess is that Dart follows those symlinks, which makes sense. I'll investigate a little more to try to understand better how things unfold.

Is there a way to make the language server extra extra verbose and record everything it does in a file?

**Comment by PaulGrandperrin** [comment]
One important thing is that, dart, my flutter sdk and all my dev env comes from this `.devenv` directory so I'm thinking that I can't really just exclude this directory.

This might also explain why the language server ignore the exclusion, because its SDK files are in there. @DanTup 

> (I'm assuming that by "project directory" you mean the directory you opened in the IDE.)

@bwilkerson Yes, that's what I meant

I'll try to give you more info later, thanks for your replies.

**Comment by DanTup** [comment]
> This might also explain why the language server ignore the exclusion, because its SDK files are in there. @DanTup 

I'm not sure there won't be issues by having your SDK inside an ignored folder, but I'm surprised that we walk all of that folder if it's excluded (in a quick test, I couldn't repro the server trying to ready anything nested inside a folder excluded with `dart.analysisExcludedFolders` but I don't know if my repro is similar enough to yours).

**Comment by PaulGrandperrin** [comment]
My bad, actually, `dart` also scans everything when only opening the flutter subproject but is somehow much faster, similarly to `rust-analyzer`.

I really don't know why, and I now need to do other things..

**Comment by DanTup** [comment]
@PaulGrandperrin are you able to reproduce this in an isolated example (eg. by creating a new project and putting a symlink inside it to somewhere further up the tree)? I wasn't able to repro the behaviour you see last time I tested, so having instructions to produce a project that triggers it would really help. Thanks!

**Comment by bwilkerson** [comment]
@PaulGrandperrin 

All of the attempts you posted above appear to be VC Code settings. Curious to know whether you've attempted to use an analysis options file to disable analysis of everything under `subProductRust`? If you're not familiar with analysis options files, or using them in this way, you can read more at https://dart.dev/tools/analysis#excluding-code-from-analysis. Id be interested to know whether adding an analysis options file at `/` allows you to open `/` without analyzing all of the files under `subProductRust` or any of the `.devenv` directories.

> One important thing is that, dart, my flutter sdk and all my dev env comes from this `.devenv` directory so I'm thinking that I can't really just exclude this directory.

Excluding files and directories in the analysis options file doesn't prevent the analyzer from looking in them when there's a direct import that it needs to follow, it just prevents it from traversing those directories to find files to analyze. It should be fine to exclude them.

**Comment by PaulGrandperrin** [comment]
@DanTup  @bwilkerson  Sorry I didn't follow up, this bug completely prevented me from continuing to work with dart/flutter so I moved on to another project.
I now have to move back to continuing this project.

I now have 32GB of RAM and `dartaotruntime` still crashes my computer but I think not by exhausting my RAM but by keeping too many files open (inotify?): after about 5 minutes of `dart analyze` running, my filesystem ZFS's `arc_prune` kernel process starts using all CPU and then system freezes and crashes, but there's still some RAM available.

Because my project is complex, I decided to start from scratch a new project and check what's the minimum I need to do to trigger the bug:


### Reproduction steps
```sh
flutter create dart_bug_52947
cd dart_bug_52947
mkdir do_not_go
ln -s /nix/store do_not_go/symlink # creates the symlink to huge directory tree, here my nix store
cat >> analysis_options.yaml<<EOF
analyzer:
  exclude:
    - do_not_go
    - do_not_go/
    - do_not_go/* 
    - do_not_go/**
EOF
dart analyze
```

Without the symlink `dart analyze` takes less than a second but with it, it takes about a minute. For some reason, in this case, the analysis actually ends and without crashing my system but it shows that `dart analyse` will follow symlinks in forbidden places.

**Comment by DanTup** [comment]
@PaulGrandperrin there are some known issues with the performance of the watcher with symlinks. @davidmorgan has been working on some fixes, but I'm not sure if they've been released yet.

Edit: Some related issues:

- https://github.com/dart-lang/sdk/issues/62008
- https://github.com/dart-lang/sdk/issues/61407
- https://github.com/dart-lang/tools/issues/2170
- https://github.com/dart-lang/tools/issues/1728

**Comment by PaulGrandperrin** [comment]
@DanTup Oh nice!
I just went back to this project and didn't update flutter yet but the 2 closed bug reports you shared seems very promising!
I'll update that asap, thanks :)

**Comment by davidmorgan** [comment]
Thanks Danny.

The `package:watcher` change that should address issues with symlinks will be in 3.11, if you're able to try a beta release, it's in 3.11.0-200.1.beta

https://dart.dev/get-dart/archive

**Comment by davidmorgan** [comment]
I guess you probably want Flutter not just the Dart SDK; looks like Flutter 3.40.0-0.1.pre beta has the Dart SDK 3.11.0-200.1.beta.

**Comment by PaulGrandperrin** [comment]
Thanks you both for your quick answers!
@davidmorgan ok, I'll try that and report back!

**Comment by PaulGrandperrin** [comment]
@DanTup @davidmorgan 
I think I'm very unlucky with Dart because I lost lots of time with what I think is probably another bug (maybe from vscode):
When I update `FLUTTER_ROOT`, `DART_ROOT` and `PATH` to point to the new flutter SDK, vscode ignores all that, and instead uses the previous SDK. It tooks me a long time to understand how it could know where the previous SDK was! It is stored in `.dart_tool/package_config.json`.
I don't know why outdated information is kept here, why it is used instead of the user specified one, and how it is supposed to be updated, but deleting the file solved the issue (which recreated the file with new SDK's path hardcoded in it).

Anyway... It doesn't work...

It definitively did something because for instance now `dart analyse` finishes but when using vscode LSP, it stills takes down my computer after a few minutes.

I think the responsible is this process:
```
/home/paulg/Repos/flutter/bin/cache/dart-sdk/bin/dartvm --resolved_executable_name=/home/paulg/Repos/flutter/bin/cache/dart-sdk/bin/dart --executable_name=/home/paulg/Repos/flutter/bin/cache/dart-sdk/bin/dart --packages=/home/paulg/Repos/flutter/packages/flutter_tools/.dart_tool/package_config.json /home/paulg/Repos/flutter/bin/cache/flutter_tools.snapshot widget-preview start --machine --web-server --dtd-url ws://127.0.0.1:33061/P6LaXtzeYjM= --devtools-server-address http://127.0.0.1:9100/
```

It's using 100% of one CPU core until the filesystem starts running out of some resource and takes down the OS.

**Comment by mraleph** [comment]
@PaulGrandperrin you should file a separate issue about that in Flutter repo. It's Flutter Widget Preview process.

**Comment by PaulGrandperrin** [comment]
I'll do that!
I think this bug report should be close, thanks for your help :)

**Comment by DanTup** [comment]
@PaulGrandperrin 

> When I update `FLUTTER_ROOT`, `DART_ROOT` and `PATH` to point to the new flutter SDK, vscode ignores all that, and instead uses the previous SDK. It tooks me a long time to understand how it could know where the previous SDK was! It is stored in `.dart_tool/package_config.json`.
> I don't know why outdated information is kept here, why it is used instead of the user specified one, and how it is supposed to be updated, but deleting the file solved the issue (which recreated the file with new SDK's path hardcoded in it).

VS Code will use `PATH` to find an SDK if it doesn't find a better one another way. It can't always use `PATH` because you may have different projects using different SDK versions. The best way to change between SDKs is with the `dart.sdkPath`/`dart.flutterSdkPath` settings [or the UI that lets you quickly switch between values](https://dartcode.org/docs/quickly-switching-between-sdk-versions/).

`.dart_tool/package_config.json` is where the resolved package information is stored, and what the Dart analysis server will use to resolve packages. VS Code will use this path if there isn't one configured (via `dart.flutterSdkPath`) to ensure that the tooling and debug sessions spawned by VS Code use the same version of the Flutter SDK as the analyzer. Otherwise things can get confusing (for example the analyzer tells you there are no errors but when you try to run your project we fail to launch with compile errors).

If you run `flutter pub get` in your project using the Flutter SDK you want to use, that would update `.dart_tool/package_config.json` (and then, VS Code would also use that version). If you have more than one SDK though, I'd recommend using the SDK switcher mentioned above because it will do everything for you and makes it easy to switch.

Unfortunately these details are not always obvious. If you think there are ways this could be improved, please do file issues at https://github.com/Dart-Code/Dart-Code!

(I'll respond on the other issue about the widget preview)

**Comment by PaulGrandperrin** [comment]
@DanTup Thanks for taking the time to explain all of this, of course there had to be a sensible technical explanation.

So I guess it's a good design for people managing SDKs the regular way (`flutter updgrade` / IDE) as it ensures everything stays in sync.

In my case, having an external system (nix) manage the SDK, it leads unsynchronized versions between the command line and the IDE.

I don't know what can and should be done about it, but I'll open an issue because in my use case, it's a very frustrating issue.

Anyway I think you can close this issue if you wish to.

**Comment by keertip** [comment]
The performance bottleneck during initial analysis and watcher startup on projects with symlinks was addressed in `package:watcher` updates included in Dart SDK 3.11 (and Flutter 3.40). The reporter confirmed the resolution with the update.

Closing this issue as resolved. If there is more work to do here, please let us know by filing a new issue with up to date information. Thanks!
